### PR TITLE
Mergin/download project

### LINKED ISSUE #119: [Mergin] Download a project

### PR DESCRIPTION
**Pull request by sklencar**
Downloading a project from Mergin.
Behavior after downloading a project has not been discussed yet, so after download, a project is not loaded, just added into "My projects" list. Downloading is visualise with busyIndicator, no other indications (e.g. project is already downloaded, downloaded project is out of date).
Added cache file for listed project.

closes #119

### CODE REVIEW

**Comment by PeterPetrik** [comment]
@sklencar "This branch cannot be rebased due to conflicts" can you please rebase?

File: app/main.cpp [hunk 1]
```diff
@@ -271,7 +271,7 @@ int main(int argc, char *argv[])
 #ifdef MERGIN_TOKEN
   merginToken.append(STR(MERGIN_TOKEN));
 #endif
-  MerginApi* ma=  new MerginApi(QString("https://mergin.dev.cloudmappin.com"), merginToken );
+  MerginApi* ma=  new MerginApi(QString("https://mergin.dev.cloudmappin.com"), dataDir, merginToken );
```
Comment (PeterPetrik) [inline app/main.cpp]: best to use smart pointers

File: app/merginapi.cpp [hunk 2]
```diff
@@ -1,33 +1,69 @@
 #include "merginapi.h"
+#include "qgsziputils.h"
```
Comment (PeterPetrik) [inline app/merginapi.cpp]: would we need a new SDK for building  with your patch?

File: app/merginapi.cpp [hunk 3]
```diff
@@ -42,18 +78,70 @@ void MerginApi::listProjectsReplyFinished()
...
+    }
+
+    return path;
+}
+
+void MerginApi::downloadProjectReplyFinished()
+{
+
+    QNetworkReply* r = qobject_cast<QNetworkReply*>(sender());
+    Q_ASSERT(r);
```
Comment (PeterPetrik) [inline app/merginapi.cpp]: just return?

File: app/merginprojectmodel.h [hunk 4]
```diff
@@ -35,6 +36,8 @@ class MerginProjectModel: public QAbstractListModel
 signals:
     void merginProjectsChanged();
 
+public slots:
+    void downloadProjectFinished(QString projectFolder, QString projectName);
   private:
     MerginApi* mApi;
```
Comment (PeterPetrik) [inline app/merginprojectmodel.h]: ufff these raw pointers looks scary all around!

File: app/projectsmodel.cpp [hunk 5]
```diff
@@ -30,7 +30,12 @@ ProjectModel::ProjectModel(const QString &dataDir, QObject* parent)
 ProjectModel::~ProjectModel() {}
 
 void ProjectModel::findProjectFiles() {
-    QDirIterator it(mDataDir, QStringList() << "*.qgs", QDir::Files, QDirIterator::Subdirectories);
+    addProjectsFromPath(mDataDir);
+}
+
+void ProjectModel::addProjectsFromPath(QString path)
+{
+    QDirIterator it(path, QStringList() << "*.qgs", QDir::Files, QDirIterator::Subdirectories);
```
Comment (PeterPetrik) [inline app/projectsmodel.cpp]: qstringliteral

File: app/merginapi.cpp [hunk 6]
```diff
@@ -42,18 +78,70 @@ void MerginApi::listProjectsReplyFinished()
...
+    QDir dir;
+    if (!dir.exists(mDataDir))
+        dir.mkpath(mDataDir);
+    QFile file(mDataDir + projectName);
+    if (file.exists()) {
+        qDebug("Zip file already exists!");
+        // TODO overwrite projects / handle data-sync
+    }
+
+    bool isOpen = file.open(QIODevice::WriteOnly);
```
[the remark this answers is not in the document]
Comment (sklencar) [reply app/merginapi.cpp]: Data is a QByteStream containing zip file (data == const QByteArray &data). Shouldn't QTextStream be used for text only? Or rather to use QDataStream if file.write is unacceptable?

**Comment by sklencar** [comment]
Closing since it has been replaced with data-stream download.

**Pull request closed without merging**